### PR TITLE
Add missing Maskformer dataclass decorator, add dataclass check in ModelOutput for subclasses

### PR DESCRIPTION
**Pull request by rachthree**
# What does this PR do?

Fixes # https://github.com/huggingface/transformers/issues/25504

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts 

Some additional notes:
* I tried finding a way to find all subclasses of `ModelOutput` that could be tested in one unittest to check to see if they were a `dataclass`, but this may not be possible without finding all files that use `ModelOutput` and importing each subclass. I instead opted for a check inside `ModelOutput` that happens if a subclass is instantiated.
* One problem I ran into was that `Pipeline` uses `ModelOutput` directly: https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/base.py#L933. This is not a true `dataclass`, so `dataclasses` utilities such as `is_dataclass` cannot be reliably used. The checks done in `ModelOutput.__post_init__` would not be executed because `__post_init__` is a `dataclass`-specific dunder. This makes me think there should be a separate `PipelineOutput` class that can be created from a `ModelOutput` to create a true `dataclass`, or a class method in `ModelOutput` that creates a true `dataclass` given arguments similar as ones for `OrderedDict`. However, I don't know what the potential impacts could be since I don't know all the inner workings of `transformers.`

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @amyeroberts

File: src/transformers/utils/generic.py
Comment (amyeroberts) [inline src/transformers/utils/generic.py:281]: nit: extra brackets aren't needed here 

```suggestion
            raise TypeError(
                f"{self.__module__}.{self.__class__.__name__} is not a dataclasss."
                " This is a subclass of ModelOutput and so must use the @dataclass decorator."
            )
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding! 

Just a tiny nit - otherwise good to merge!

File: src/transformers/utils/generic.py
Comment (rachthree) [reply src/transformers/utils/generic.py]: Thanks!

**Comment by rachthree** [comment]
Thank you for reviewing, @amyeroberts! How do you feel about my 2nd bullet point under the additional notes? Should I open another issue?

**Comment by amyeroberts** [comment]
@rachthree For the second point, is it necessary that the checks in `__post_init__` are run within the pipeline code? Providing our model output classes inherit from `ModelOutput`, the checks will be performed when the classes are used, which I think is sufficient. We don't want to add an additional output class in this case - pipelines are essentially just glue which holds together the preprocessing, inference and postprocessing steps, so should handle the same objects. 

I'm not sure I got all the implications of your question however. If you think I've missed the point, let me know.

**Comment by rachthree** [comment]
@amyeroberts My apologies for my late response, I was on travel for a bit. Could the workflow for my PR be approved? CI got triggered again once I pushed the commit. Thanks!

Regarding dataclasses, the checks are not necessary if those checks are specifically for a model's subclassed `ModelOutput` with the `@dataclass` decorator, but the reason why I'm suggesting an additional output class is that `output = ModelOutput(foo_inputs)` by itself is not a true `dataclass`. A new class that is a true `dataclass` can be created using `dataclasses.makedataclass`, perhaps in some new classmethod for `ModelOutput`. I originally was thinking something like `PipelineOutput.from_dict` if using a new class.

Should anyone try to use `dataclass` utilities such as `dataclass.fields` on a Pipeline's `ModelOutput`, those utilities will not function properly:

```python
from transformers.utils.generic import ModelOutput
import dataclasses

output = ModelOutput({"a": 1, "b": 2})

fields = dataclasses.fields(output)
```

Results in `TypeError: must be called with a dataclass type or instance`

**Comment by amyeroberts** [comment]
@rachthree Let's just stick to the changes here in this PR for now. If pipeline consistently use `ModelOutput` directly, then it shouldn't cause an issue. If it does. then we can open a new, separate PR to address it.

**Comment by amyeroberts** [comment]
@rachthree Let me know if there's any other changes you'd like to push before I merge

**Comment by rachthree** [comment]
@amyeroberts Totally fair, and didn't meant to add more dataclass-related changes to this PR. Leaving it for now sounds good to me. No more changes on my end, thanks!

**Comment by lixiaoqingnnz** [comment]
@amyeroberts 
Hi, I found the same problem in the model: BiomedVLP-CXR-BERT-specialized.

reproduce (code found in https://huggingface.co/microsoft/BiomedVLP-CXR-BERT-specialized):
import torch
from transformers import AutoModel, AutoTokenizer

# Load the model and tokenizer
url = "microsoft/BiomedVLP-CXR-BERT-specialized"
tokenizer = AutoTokenizer.from_pretrained(url, trust_remote_code=True)
model = AutoModel.from_pretrained(url, trust_remote_code=True)

# Input text prompts (e.g., reference, synonym, contradiction)
text_prompts = ["There is no pneumothorax or pleural effusion",
                "No pleural effusion or pneumothorax is seen",
                "The extent of the pleural effusion is constant."]

# Tokenize and compute the sentence embeddings
tokenizer_output = tokenizer.batch_encode_plus(batch_text_or_text_pairs=text_prompts,
                                               add_special_tokens=True,
                                               padding='longest',
                                               return_tensors='pt')
embeddings = model.get_projected_text_embeddings(input_ids=tokenizer_output.input_ids,
                                                 attention_mask=tokenizer_output.attention_mask)

# Compute the cosine similarity of sentence embeddings obtained from input text prompts.
sim = torch.mm(embeddings, embeddings.t())

Error that I encountered:
TypeError: transformers_modules.microsoft.BiomedVLP-CXR-BERT-specialized.b59c09e51ab2410b24f4be214bbb49043fe63fc2.modeling_cxrbert.CXRBertOutput is not a dataclasss. This is a subclass of ModelOutput and so must use the @dataclass decorator.


Could you add the missing datacalss decorator to this model also?
Thank you so much!

**Comment by lixiaoqingnnz** [comment]
@amyeroberts Thanks for your reply, I already added the decorated and the problem have been solved. Sorry that I bothered you cause I thought this repo maintainers is also you. 
Thanks anyway again!

**Comment by amyeroberts** [comment]
@lixiaoqingnnz No worries - it wasn't a bother :)

**Pull request merged**
